### ISSUE TITLE
tests/drivers/watchdog/wdt_basic_api fails on Quark SE / x86 and esp32

### ISSUE THREAD

**Issue by inakypg** [zephyrproject-rtos/zephyr#8097]
Toolchain: ZEPHYR_GCC_VARIANT=zephyr

Commit 43553da9b2bb9e537185695899fb5184c3b17ebe

Testcase seems to never complete after restarting the system

**Error Console Log on quark_se_c1000/x86:**
```
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS 1.12.0-rc2 (delayed boot 1000ms) *****
Running test suite wdt_basic_test
===================================================================
starting test - test_wdt
Testcase: test_wdt_no_callback
***** delaying boot 1000ms (per build configuration) *****
***** Booting Zephyr OS 1.12.0-rc2 (delayed boot 1000ms) *****
Running test suite wdt_basic_test
===================================================================
starting test - test_wdt
Testcase: test_wdt_no_callback
```

**Error Console Log on esp32:**
```
***** Booting Zephyr OS v1.12.0-831-g5b8e4ae (delayed boot 1000ms) *****
Running test suite wdt_basic_test
===================================================================
starting test - test_wdt
Testcase: test_wdt_no_callback
 ** FATAL EXCEPTION
 ** CPU 0 EXCCAUSE 20 PS 0x00060120 PC 0x00000000 VADDR 0x00000000
 **  A0 0x80082419  SP 0x3ffb1b70  A2 0x00000000  A3 0x00000000
 **  A4 0x3ffb0000  A5 0x3ffb0208  A6 0x3ffb0208  A7 0x00000000
 **  A8 0x8008099f  A9 0x3ffb1b20 A10 0x3ffb0000 A11 0x3ffb1280
 ** A12 0x3ffb036f A13 0x3ffb26e0 A14 0x3ffb021c A15 0x3ffb14c0
 ** LBEG 0x00000000 LEND 0x00000000 LCOUNT 0x00000000
 ** SAR 0x00000000
Current thread ID = 0x3ffb021c
Faulting instruction address = 0xdeaddead
Fatal fault in ISR! Spinning...
```

**Comment by tbursztyka** [comment]
I don't know why this was assigned to me, but running a git bisect (I put the good point on 1.11) points to commit 7a5ff13703a0139f917afd58151934b5fcf19bb9

@nashif Can you check?

That said, while running the git bisect, I realized that test (which is made only for qmsi... that sucks, moreover it's supposed to test the generic API but anyway) gives a slight different output on success sometimes: sometime it resets and pass, sometime it does not reset but still passes.

**Comment by nashif** [comment]
r

> @nashif Can you check?

i just hit this commit with bisect myself, looking

**Comment by nashif** [comment]
@punitvara  status update please? any fix in sight?

**Comment by ManojSubbarao** [comment]
@punitvara Please take a quick look and drive this issue to closure. This is pending from long time without any updates.

**Comment by SavinayDharmappa** [comment]
@nashif  I did some intial analysis, the watch dog timer for qmsi and esp32 are not moved to new interfaces introduced . But application uses newly introduced interface for which the QMSI and esp32 WDT driver are not compaitable. This reason why esp32 crashes. Still wondering why null pointer derefercing of QMSI driver is not showing crash log instead just hangs.

**Comment by markusr** [comment]
I also have a problem with watchdog example on STM32 (```nucleo_l432kc```). 

```
***** Booting Zephyr OS v1.13.0-rc2-1-g19b7eaa *****
***** USAGE FAULT *****
  Illegal use of the EPSR
***** Hardware exception *****
Current thread ID = 0x2000022c
Faulting instruction address = 0x80200288
Fatal fault in essential thread! Spinning...
***** Booting Zephyr OS v1.13.0-rc2-1-g19b7eaa *****
***** USAGE FAULT *****
  Illegal use of the EPSR
***** Hardware exception *****
Current thread ID = 0x2000022c
Faulting instruction address = 0x80200288
Fatal fault in essential thread! Spinning...
```

**Comment by SavinayDharmappa** [comment]
@nashif @ManojSubbarao  Modified wdt driver to new interfaces defined in watchdog.h. As of now only no callback test case is passing other two test cases are failing. I am still debugging the issue.

**Comment by SavinayDharmappa** [comment]
After wdt qmsi driver changes , test cases for watchdog are passing. I have tested it on quark_se_c1000 board. I will work on esp32 driver changes as this driver also needs to be migrated for new wdt API interfaces.

**Comment by SavinayDharmappa** [comment]
Implemented new wdt api interface for esp32. After that test case are failing, because global data are re-initialized upon system reset due watchdog time expiry. I am still debugging. This not the case for quark_se_c100.

**Comment by SavinayDharmappa** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/9872

**Comment by ManojSubbarao** [comment]
@SavinayDharmappa  please look into this issue

**Comment by SavinayDharmappa** [comment]
This issue is reproduced only in quark d2000 and arc of quark_se_c1000 . The driver changes for x86 quark_se_c1000 part  still debugging to find the root cause.

**Comment by SavinayDharmappa** [comment]
In case of quark_d2000 watchdog reset is triggered , but variable defined in the app_noinit section also getting reinitialized during watchdog reset, which not the case for other platforms. Because of which "test_wdt_no_callback" repeats indefinitely.

**Comment by SavinayDharmappa** [comment]
After further debugging it clear from status register reset is triggered by watchdog only but ram contents are being lost in that reset. Because of which test is going repeated loop. I think we have bypass this testcase for quark_d2000 @nashif

**Comment by SavinayDharmappa** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/13659

**Comment by ManojSubbarao** [comment]
Closed via #13659
